includes/class-wc-subscriptions-data-copier.php

```diff
@@ -0,0 +1,461 @@
+		'_order_key',
+		'_edit_lock',
+		'_wc_points_earned',
+		'_transaction_id',
+		'_billing_interval',
+		'_billing_period',
+		'_subscription_resubscribe',
+		'_subscription_renewal',
+		'_subscription_switch',
+		'_payment_method',
+		'_payment_method_title',
+		'_suspension_count',
+		'_requires_manual_renewal',
+		'_cancelled_email_sent',
+		'_trial_period',
+		'_created_via',
+		'_order_stock_reduced',
+	];
+
+	/**
+	 * The subscription or order being copied.
+	 *
+	 * @var WC_Order
+	 */
+	private $from_object = null;
+
+	/**
+	 * The subscription or order being copied to.
+	 *
+	 * @var WC_Order
+	 */
+	private $to_object = null;
+
+	/**
+	 * The type of copy. Can be 'subscription' or 'renewal'.
+	 *
+	 * Used in dynamic filters to allow third parties to target specific meta keys in different copying contexts.
+	 *
+	 * @var string
+	 */
+	private $copy_type = '';
+
+	/**
+	 * Copies data from one object to another.
+	 *
+	 * This function acts as a publicly accessible wrapper for obtaining an instance of the copier and completing the copy.
+	 *
+	 * @param WC_Order $from_object The object to copy data from.
+	 * @param WC_Order $to_object   The object to copy data to.
+	 * @param string   $copy_type   Optional. The type of copy. Can be 'subscription', 'parent', 'renewal_order' or 'resubscribe_order'. Default is 'subscription'.
+	 */
+	public static function copy( $from_object, $to_object, $copy_type = self::DEFAULT_COPY_TYPE ) {
+		$instance = new self( $from_object, $to_object, $copy_type );
+		$instance->copy_data();
+	}
+
+	/**
+	 * Constructor.
+	 *
+	 * @param WC_Order $from_object The object to copy data from.
+	 * @param WC_Order $to_object   The object to copy data to.
+	 * @param string   $copy_type   Optional. The type of copy. Can be 'subscription', 'parent', 'renewal_order' or 'resubscribe_order'. Default is 'subscription'.
+	 */
+	public function __construct( $from_object, $to_object, $copy_type = self::DEFAULT_COPY_TYPE ) {
+		$this->from_object = $from_object;
+		$this->to_object   = $to_object;
+		$this->copy_type   = $copy_type;
+	}
+
+	/**
+	 * Copies the data from the "from" object to the "to" object.
+	 */
+	public function copy_data() {
+
+		if ( $this->is_using_cpt_data_stores() ) {
+			$data_array = $GLOBALS['wpdb']->get_results( $this->get_deprecated_meta_query(), ARRAY_A );
+			$data       = wp_list_pluck( $data_array, 'meta_value', 'meta_key' );
+		} else {
+			$data  = $this->get_meta_data();
+			$data += $this->get_order_data();
+			$data += $this->get_operational_data();
+			$data += $this->get_address_data();
+
+			// Remove any excluded meta keys.
+			$data = $this->filter_excluded_meta_keys_via_query( $data );
+		}
+
+		$data = $this->apply_deprecated_filter( $data );
+		$data = apply_filters( "wc_subscriptions_{$this->copy_type}_data", $data, $this->to_object, $this->from_object );
+		$data = apply_filters( 'wc_subscriptions_copied_data', $data, $this->to_object, $this->from_object, $this->copy_type );
```

Comment:
What are our thoughts on these new filters?

The `wc_subscriptions_copied_data` one is a little lacking imo. I originally wanted to introduce a catch all filter so 3PDs don't have to hook onto all the different types if they just want to remove their meta from all types. 
The filter name just seems a little strange - it is the data copied from one object to another object but in the past the filters were `wcs_subscription_meta`, `wcs_parent_meta`, `wcs_renewal_order_meta`, `wcs_resubscribe_order_meta`.


`wc_subscriptions_{$this->copy_type}_data` also on this one, the actual hooks will be:

- `wc_subscriptions_subscription_data`
- `wc_subscriptions_parent_data`
- `wc_subscriptions_renewal_order_data`
- `wc_subscriptions_resubscribe_order_data`

Reply:
Some more options:
- `wc_subscriptions_copy_object_data`
- `wc_subscriptions_object_data_to_copy` - saying "copied" implies it's been copied at the time of this filter so I was thinking maybe implying this data is still to be copied.
- `wc_subscriptions_object_data` - This suggestion is inline with the current hooks i.e. "wc_subscriptions_renewal_order_data" but with a generic type instead of dynamically setting the type in the hook, it kind of feels wrong not having some wording related to copy in it

Reply:
I like the `wc_subscriptions_object_data` option. I've changed it to that in [b8cbbf0](https://github.com/Automattic/woocommerce-subscriptions-core/pull/229/commits/b8cbbf052a135dcde532c368127c763a50bbbb9f)